File: Sorting/Ruby/QuickSort.rb
def swap(arr, i, j)
    temp = arr[i]
    arr[i] = arr[j]
    arr[j] = temp
    return arr
end # end func

def quickSort(arr)
    return qsort(arr,0,arr.length-1)
end # end func

def qsort(arr, low, high)
    if low < high then
        pivot = partition(arr, low, high)
        qsort(arr, low, pivot-1)
        qsort(arr, pivot+1, high)
    end # end if
    return arr
end # end func

def partition(arr, low, high)
    i = (low -1)
    pivot = arr[high]
    for j in (low..high-1) do
        if arr[j] <= pivot then
            i += 1
            arr = swap(arr, i, j)
        end # end if
    end # end for
    arr = swap(arr, i+1, high)
    return (i+1)
end # end func

File: Sorting/Java/QuickSort.java

public class QuickSort implements SortingAlgorithm {
	protected void swap(int[] a, int i, int j){
		int temp = a[i];
		a[i] = a[j];
		a[j] = temp;
	}
	
	protected void qsort(int[] a, int top, int bot){
		int pivot = patition(a, top, bot);
		if (top<pivot-1)
			qsort(a,top,pivot-1); // Base case
		if (bot>pivot)
			qsort(a,pivot,bot); // Base case
	}
	
	protected int patition(int[] a, int top, int bot){
		int i = top;
		int j = bot;
		int pivot = (top+bot)/2;
		
		while (i<=j){
			while (a[i]<a[pivot])
				i++;
			while (a[j]>a[pivot])
				j--;
			if (i<= j){
				if (i == pivot)
					pivot = j;
				else if (j == pivot)
					pivot = i;
				swap(a,i++,j--);
			}
		}
		
		return i;
	}
	
	
	public void sort(int [] a){
		qsort(a,0,a.length-1);
	}
}
